builtin/providers/azure/resource_azure_disk.go

```diff
@@ -0,0 +1,301 @@
+	mc := meta.(management.Client)
+	lun := d.Get("lun").(int)
+	vm := d.Get("virtual_machine").(string)
+
+	if d.HasChange("size") {
+		p := virtualmachinedisk.UpdateDiskParameters{
+			Name: d.Id(),
+		}
+
+		if d.HasChange("size") {
+			p.ResizedSizeInGB = d.Get("size").(int)
+		}
+
+		log.Printf("[DEBUG] Updating disk: %s", d.Id())
+		req, err := virtualmachinedisk.NewClient(mc).UpdateDisk(d.Id(), p)
+		if err != nil {
+			return fmt.Errorf("Error updating disk %s: %s", d.Id(), err)
+		}
+
+		// Wait until the disk is updated
+		if err := mc.WaitForOperation(req, nil); err != nil {
+			return fmt.Errorf(
+				"Error waiting for disk %s to be updated: %s", d.Id(), err)
+		}
+	}
+
+	if d.HasChange("virtual_machine") {
+		log.Printf("[DEBUG] Detaching data disk: %s", d.Id())
+		req, err := virtualmachinedisk.NewClient(mc).DeleteDataDisk(vm, vm, vm, lun, false)
+		if err != nil {
+			return fmt.Errorf("Error detaching data disk %s: %s", d.Id(), err)
+		}
+
+		// Wait until the data disk is detached
+		if err := mc.WaitForOperation(req, nil); err != nil {
+			return fmt.Errorf(
+				"Error waiting for data disk %s to be detached: %s", d.Id(), err)
+		}
+
+		p := virtualmachinedisk.CreateDataDiskParameters{
+			DiskName:    d.Id(),
+			Lun:         lun,
+			HostCaching: hostCaching(d),
+			MediaLink:   mediaLink(d),
+		}
+
+		log.Printf("[DEBUG] Attaching data disk: %s", d.Id())
+		req, err = virtualmachinedisk.NewClient(mc).AddDataDisk(vm, vm, vm, p)
+		if err != nil {
+			return fmt.Errorf("Error attaching data disk %s to instance %s: %s", d.Id(), vm, err)
+		}
+
+		// Wait until the data disk is attached
+		if err := mc.WaitForOperation(req, nil); err != nil {
+			return fmt.Errorf(
+				"Error waiting for data disk %s to be attached to instance %s: %s", d.Id(), vm, err)
+		}
+
+		// Make sure we return here since all possible changes are
+		// already updated if we reach this point
+		return nil
+	}
+
+	if d.HasChange("caching") || d.HasChange("lun") {
+		p := virtualmachinedisk.UpdateDataDiskParameters{
+			DiskName:    d.Id(),
+			Lun:         lun,
+			HostCaching: hostCaching(d),
+			MediaLink:   mediaLink(d),
+		}
+
+		log.Printf("[DEBUG] Updating data disk: %s", d.Id())
+		req, err := virtualmachinedisk.NewClient(mc).UpdateDataDisk(vm, vm, vm, lun, p)
+		if err != nil {
+			return fmt.Errorf("Error updating data disk %s: %s", d.Id(), err)
+		}
+
+		// Wait until the data disk is updated
+		if err := mc.WaitForOperation(req, nil); err != nil {
+			return fmt.Errorf(
+				"Error waiting for data disk %s to be updated: %s", d.Id(), err)
+		}
+	}
+
+	return resourceAzureDataDiskRead(d, meta)
+}
+
+func resourceAzureDataDiskDelete(d *schema.ResourceData, meta interface{}) error {
+	mc := meta.(management.Client)
+	lun := d.Get("lun").(int)
+	vm := d.Get("virtual_machine").(string)
+
+	// If a name was not supplied, it means we created a new emtpy disk and we now want to
```

Comment:
Your logic here is pretty hard to follow; the feature of an actively chosen delete/detach option is a great idea though.
Personally; I would make a separate parameter called ephemeral or whatever and judge the disk's deletion/detachment according to that one.